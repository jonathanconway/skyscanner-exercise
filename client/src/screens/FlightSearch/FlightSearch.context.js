import { createContext } from 'react';

const FlightSearchContext = createContext({ });

export default FlightSearchContext;
