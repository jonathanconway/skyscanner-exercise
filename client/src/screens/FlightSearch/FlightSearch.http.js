import { http as httpUtils } from '../../utils';

const getSearch = async (query) => {
  const host = 'http://localhost:4000',
        path = '/api/search',
        queryString = httpUtils.encodeQuery(query);

  const response = await fetch(`${host}${path}?${queryString}`);

  if (!response.ok) {
    throw response.statusText;
  }

  return response.json();
};

const http = {
  getSearch
};

export default http;