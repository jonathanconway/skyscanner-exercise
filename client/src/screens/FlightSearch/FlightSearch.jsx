import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { BpkExtraLargeSpinner } from 'bpk-component-spinner';
import BpkBannerAlert, { ALERT_TYPES } from 'bpk-component-banner-alert';

import FlightSearchContext from './FlightSearch.context';
import Header from './Header/Header';
import Actions from './Actions/Actions';
import Itinerary from './Itinerary/Itinerary';

import STYLES from './FlightSearch.scss';
import http from './FlightSearch.http';

const c = className => STYLES[className] || 'UNKNOWN';

export const DEFAULT_TAKE = 5;

const FlightSearch = (props) => {
  const {
          originAirportCode,
          destinationAirportCode,
          outboundDate,
          returnDate,
          currencyCode
        } = props,

        [isLoadingItineraries, setIsLoadingItineraries] = useState(false),
        [error, setError] = useState(undefined),
        [itineraries, setItineraries] = useState(undefined);

  useEffect(() => {
    const loadItineraries = async () => {
      setError(undefined);
      setIsLoadingItineraries(true);
  
      try {
        const itineraries = await http.getSearch({
            originAirportCode,
            destinationAirportCode,
            outboundDate,
            returnDate,
            take: DEFAULT_TAKE,
            currencyCode
          });

        setItineraries(itineraries);
      } catch (ex) {
        setError(ex);
      }
  
      setIsLoadingItineraries(false);
    };
    
    loadItineraries();
  }, []);

  return (
    <FlightSearchContext.Provider value={props}>
      <div className={c('FlightSearch')}>
        <Header />

        <Actions />

        <div className={c('FlightSearch__itineraries')}>
          {isLoadingItineraries &&
            <BpkExtraLargeSpinner className={c('FlightSearch__itineraries-spinner')} />}
          
          {error &&
            <BpkBannerAlert
              className={c('FlightSearch__itineraries-error')}
              message={`Sorry! There was an error: '${error}'. Please try refreshing.`}
              type={ALERT_TYPES.ERROR}
            />}

          {itineraries && itineraries.itineraries.map(itinerary => (
            <Itinerary
              key={itinerary.id}
              {...itinerary}
            />
          ))}
        </div>
      </div>
    </FlightSearchContext.Provider>
  );
};

FlightSearch.propTypes = {
  originAirportCode: PropTypes.string,
  destinationAirportCode: PropTypes.string,
  outboundDate: PropTypes.string,
  returnDate: PropTypes.string,
  numberOfTravellers: PropTypes.oneOf([1, 2]),
  currencyCode: PropTypes.string,
  flightClass: PropTypes.oneOf(['economy', 'business'])
};

export default FlightSearch;
