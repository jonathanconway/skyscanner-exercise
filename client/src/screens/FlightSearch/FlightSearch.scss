@import '~bpk-mixins';

@import '../../theme';

.FlightSearch {
  position: relative;
  display: flex;
  flex-direction: column;
  flex: 1;

  & > * {
    padding: $bpk-spacing-sm;
  }

  &__itineraries {
    flex: 1;
    overflow-y: auto;
    -webkit-overflow-scrolling: touch;

    &-spinner {
      position: absolute;
      left: 50%;
      top: 50%;
      margin-left: -($bpk-icon-size-lg / 2);
      margin-top: -($bpk-icon-size-lg / 2);
      fill: $brand-color;
      z-index: 1;
    }

    &-next-button {
      @extend .action-button;
      
      position: relative;
      display: inline-block;
      margin-top: auto;
      width: 100%;
      height: 3rem;

      &-spinner {
        position: absolute;
        left: 50%;
        top: 50%;
        margin-left: -($bpk-icon-size-sm / 2);
        margin-top: -($bpk-icon-size-sm / 2);
        width: $bpk-icon-size-sm;
        height: $bpk-icon-size-sm;
        fill: black !important;
      }
    }
  }
}
