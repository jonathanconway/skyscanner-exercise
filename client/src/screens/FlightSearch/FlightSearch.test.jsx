import React from 'react';
import ReactDOM from 'react-dom';
import { act } from "react-dom/test-utils";
import ShallowRenderer from 'react-test-renderer/shallow';

import http from "./FlightSearch.http";
import FlightSearch, { DEFAULT_TAKE } from './FlightSearch';

describe('FlightSearch', () => {
  it('should render without crashing', () => {
    const div = document.createElement('div');

    act(() => {
      ReactDOM.render(<FlightSearch />, div);
    });
  });

  it('should render correctly', () => {
    const renderer = new ShallowRenderer();

    act(() => {
      renderer.render(<FlightSearch />);
    });

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  describe('on mount behavior', () => {
    const div = document.createElement('div');
  
    jest.spyOn(http, 'getSearch')
      .mockImplementationOnce(async () => { })
      .mockImplementationOnce(async () => { })
      .mockImplementationOnce(async () => { throw Error('MESSAGE'); });

    const flightSearchNode =
      <FlightSearch
        originAirportCode='EDI'
        destinationAirportCode='LHR'
        outboundDate='2019-01-01'
        numberOfTravellers={2}
        currencyCode='GBP'
        flightClass="economy"
      />;

    act(() => {
      ReactDOM.render(flightSearchNode, div);

      it('should perform a flight search on mount', (done) => {
        setTimeout(() => {
          expect(http.getSearch).toHaveBeenCalledWith({
            originAirportCode: 'EDI',
            destinationAirportCode: 'LHR',
            outboundDate: '2019-01-01',
            take: DEFAULT_TAKE,
            currencyCode: 'GBP'
          });
          done();
        });
      });
    });
  
    it('should show spinner during loading and then hide it after loading', (done) => {
      const isSpinnerVisible = () => div.getElementsByClassName('FlightSearch__itineraries-spinner').length > 0;

      expect(isSpinnerVisible()).toBeFalsy();
  
      setTimeout(() => {
        expect(isSpinnerVisible()).toBeFalsy();
        done();
      });
    });
  
    it('should show an error if flight search fails', (done) => {
      const isErrorMessageVisible = () => div.getElementsByClassName('FlightSearch__itineraries-error').length > 0;

      expect(isErrorMessageVisible()).toBeFalsy();
  
      setTimeout(() => {
        expect(isErrorMessageVisible()).toBeTruthy();
        done();
      });
    });
  });
});

// describe('FlightSearch.http', () => {
//   it('should call fetch with appropriate querystring parameters', async () => {
//     const mockJson = jest.fn(async () => ({}));
//     const fetchSpy = jest.spyOn(global, 'fetch');
//     fetchSpy.mockImplementation(() => Promise.resolve({ ok: true, json: mockJson })); 

//     await http.getSearch({
//       originAirportCode: 'EDI',
//       destinationAirportCode: 'LHR',
//       outboundDate: '2019-01-01',
//       currencyCode: 'GBP',
//       take: 5
//     });

//     expect(fetchSpy).toHaveBeenCalledWith(expect.stringContaining('originAirportCode=EDI'));
//     expect(fetchSpy).toHaveBeenCalledWith(expect.stringContaining('destinationAirportCode=LHR'));
//     expect(fetchSpy).toHaveBeenCalledWith(expect.stringContaining('outboundDate=2019-01-01'));
//     expect(fetchSpy).toHaveBeenCalledWith(expect.stringContaining('currencyCode=GBP'));
//     expect(mockJson).toHaveBeenCalled();
//   });

//   it('should throw an error if status if not ok', async () => {
//     const fetchSpy = jest.spyOn(global, 'fetch');
//     fetchSpy.mockImplementation(() => Promise.resolve({ ok: false })); 

//     let exception = false;
//     try {
//       await http.getSearch({});
//     } catch (ex) {
//       exception = true;
//     }
//     expect(exception).toBeTruthy();
//   });
// });
