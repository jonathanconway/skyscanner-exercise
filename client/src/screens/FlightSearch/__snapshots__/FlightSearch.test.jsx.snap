// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlightSearch should render correctly 1`] = `
<Context.Provider
  value={Object {}}
>
  <div
    className="FlightSearch"
  >
    <Header />
    <Actions />
    <div
      className="FlightSearch__itineraries"
    />
  </div>
</Context.Provider>
`;
